### PR TITLE
Implement `Selector` (#574 part3)

### PR DESCRIPTION
**Pull request by inikulin**
In this part:

- Node existence awaiting
- `clientFunction.with(options)` instead of `.bindTestRun`
- `selector.with({ visibilityCheck: Boolean, timeout: Number})`
- Removed `t.waitForElement`
- RAW API now handles errors the same way ESNext API does
- Actions now use selectors machinery
- Refactored command construction and validation

\cc @helen-dikareva @AlexanderMoskovkin @VasilyStrelyaev

### CODE REVIEW

File: src/client-functions/client-function-factory.js [hunk 1]
```diff
@@ -46,46 +48,65 @@ export default class ClientFunctionFactory {
 
         var factory = this;
 
-        clientFn.bindTestRun = function bindTestRun (t) {
-            // NOTE: we can't use strict `t instanceof TestController`
-            // check due to module circular reference
-            if (!t || !(t.testRun instanceof TestRun))
-                throw new APIError('bindTestRun', MESSAGE.invalidClientFunctionTestRunBinding);
+        clientFn.with = function (options) {
```
Comment (AlexanderMoskovkin) [inline src/client-functions/client-function-factory.js]: can we avoid `var factory = this;` via

```js
clientFn.with = options => {
    this._validationOptions(options);
    return this.getFunction(options);
}
```
?

Comment (inikulin) [inline src/client-functions/client-function-factory.js]: As far as I remember in that case Babel desugaring causes problems with callsite. But I'll give it a try.

**Comment by inikulin** [comment]
@AlexanderMoskovkin Yeah, arrow worked just fine.

File: test/functional/fixtures/api/es-next/client-function/test.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ describe('[API] ClientFunction', function () {
         return runTests('./testcafe-fixtures/client-fn-test.js', 'Hammerhead code instrumentation');
     });
 
-    it('Should be able to bind a test run using `.bindTestRun(t)` method', function () {
+    it('Should be able to bind a test run using', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]: using what?

Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/client-function/test.js]:
```
    it('Should be able to bind a test run using the boundTestRun option', function () {
```
I think

File: test/functional/fixtures/api/es-next/selector/testcafe-fixtures/selector-test.js [hunk 3]
```diff
@@ -190,3 +190,50 @@ test('String ctor argument', async () => {
...
+
+test('Wait for element in DOM', async t => {
+    await t.click('#createElement');
+
+    const el = await Selector('#newElement')();
+
+    expect(el.tagName).eql('div');
+});
+
+test('Element do not appear', async () => {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/selector/testcafe-fixtures/selector-test.js]:
```
test('Element does not appear', async () => {
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by helen-dikareva** [comment]
May be better use just `validations` instead of `prop-validations` 
in e.g. ` src/test-run/commands/prop-validations/argument.js`

**Comment by inikulin** [comment]
@helen-dikareva OK

**Comment by inikulin** [comment]
FPR

File: test/functional/fixtures/api/es-next/selector/test.js [hunk 4]
```diff
@@ -25,7 +25,32 @@ describe('[API] Selector', function () {
         return runTests('./testcafe-fixtures/selector-test.js', 'String ctor argument');
     });
 
+    it('Should wait for element to appear in DOM', function () {
+        return runTests('./testcafe-fixtures/selector-test.js', 'Wait for element in DOM', { elementAvailabilityTimeout: 2500 });
+    });
+
+    it('Should return `null` or `undefined` if element does not appear within given time', function () {
```
Comment (helen-dikareva) [inline test/functional/fixtures/api/es-next/selector/test.js]: https://github.com/DevExpress/testcafe/pull/598/files#diff-f9f4a656b17573b16cd155de3e706528R205
we expect exactly `null`

**Comment by inikulin** [comment]
FPR

**Comment by helen-dikareva** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
just unstable client-tests (FYI: it seems something wrong with network connection on the farm. Most of latest unstable tests are because of too long response from the server (for iframe or xhr)).

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Pull request merged**